app/models/concerns/virtual_attribute.rb

```diff
@@ -40,10 +40,19 @@ def virtual_attribute(attribute, cast_type: :string, &block)
         _define_virtual_attribute_setter(attribute)
         _define_virtual_attribute_getter(attribute, &block)
         _define_virtual_attribute_reload(attribute)
+        _define_virtual_attributes_hook(attribute)
       end
 
       private
 
+      def _define_virtual_attributes_hook(attribute)
```

Comment:
@ulferts This has come up after the fixing of the contract, before the relation_type wasn't marked `changed?` because attributes did not return anything.

I have looked into the Attributes API and it only appears to use its own value, which is nil before the getter has been ever called. `/activerecord-5.1.6/lib/active_record/attribute.rb:36`